janusgraph-core/src/main/java/org/janusgraph/diskstorage/configuration/ExecutorServiceBuilder.java

```diff
@@ -0,0 +1,144 @@
+                " executor service keepAliveTime must be provided.");
+        }
+
+        log.info("Initiated cached thread pool of core size {}, max size {} and keep alive time {} ms",
+            corePoolSize, maxPoolSize, keepAliveTime);
+
+        if(threadFactory == null){
+            return new ThreadPoolExecutor(corePoolSize, maxPoolSize, keepAliveTime, TimeUnit.MILLISECONDS, new SynchronousQueue<>());
+        }
+
+        return new ThreadPoolExecutor(corePoolSize, maxPoolSize, keepAliveTime, TimeUnit.MILLISECONDS, new SynchronousQueue<>(), threadFactory);
+    }
+
+    private static ExecutorService buildExecutorServiceFromClassName(ExecutorServiceConfiguration executorServiceConfiguration){
+
+        String configurationClass = executorServiceConfiguration.getConfigurationClass();
+
+        Class<?> executorServiceClass;
+        try {
+            executorServiceClass = Class.forName(configurationClass);
+        } catch (ClassNotFoundException e) {
+            throw new IllegalStateException("No ExecutorService class found with class name: "+configurationClass);
+        }
+
+        if(!ExecutorService.class.isAssignableFrom(executorServiceClass)){
+            throw new IllegalStateException(configurationClass + "isn't a subclass of "+ExecutorService.class.getName());
+        }
+
+        ExecutorService executorService;
+        Constructor<?> parameterlessConstructor = null;
+        Constructor<?> executorServiceConfigurationConstructor = null;
+
+        for (Constructor<?> constructor : executorServiceClass.getDeclaredConstructors()) {
+            if (constructor.getParameterCount() == 1){
+                if(ExecutorServiceConfiguration.class.isAssignableFrom(constructor.getParameterTypes()[0])){
+                    executorServiceConfigurationConstructor = constructor;
+                    break;
+                }
+            } else if(constructor.getParameterCount() == 0){
+                parameterlessConstructor = constructor;
+            }
+        }
+
+        try {
+
+            if(executorServiceConfigurationConstructor != null){
+                executorService = (ExecutorService) executorServiceConfigurationConstructor.newInstance(executorServiceConfiguration);
+            } else {
+                if(parameterlessConstructor == null){
+                    throw new IllegalStateException(configurationClass + " has neither public constructor which accepts "
+                        +ExecutorServiceConfiguration.class.getName() + " nor parameterless public constructor.");
+                }
+                executorService = (ExecutorService) parameterlessConstructor.newInstance();
+            }
+
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalStateException("Couldn't create a new instance of "+configurationClass +
+                ". Please, check that it has a public constructor without arguments.", e);
```

Comment:
It can be reached if there is a single private parameterless constructor or single private constructor with `ExecutorService` .
See test `shouldNotCreateBackendExecutorServiceIfClassHasNoPublicConstructor`

Reply:
The above code returns up to 2 constructors. Without and with parameter. The above code doesn't check if those constructors are actually available to be accessed. Thus, the code if this `try` can actually fail with `private` constructors only.

Reply:
Got it, so the exception message here emphasizes "public" rather than "has a constructor". That makes sense.

Reply:
Yes, that's right. The exception is more about `public` here.